packages/state-transition/src/cache/epochCache.ts

```diff
@@ -500,40 +576,79 @@ export class EpochCache {
+    const nextActiveIndicesLength = epochTransitionCache.nextEpochShufflingActiveIndicesLength;
 
-    // Roll current proposers into previous proposers for metrics
+    // move current to previous
+    this.previousShuffling = this.currentShuffling;
+    this.previousDecisionRoot = this.currentDecisionRoot;
     this.proposersPrevEpoch = this.proposers;
 
-    const currentProposerSeed = getSeed(state, this.currentShuffling.epoch, DOMAIN_BEACON_PROPOSER);
-    this.proposers = computeProposers(currentProposerSeed, this.currentShuffling, this.effectiveBalanceIncrements);
+    // move next to current or calculate upcoming
+    this.currentDecisionRoot = this.nextDecisionRoot;
+    if (this.nextShuffling) {
+      // was already pulled by the api or another method on EpochCache
+      this.currentShuffling = this.nextShuffling;
+      this.shufflingCache?.metrics?.shufflingCache.nextShufflingOnEpochCache.inc();
+    } else {
+      this.currentShuffling =
+        this.shufflingCache?.getSync(upcomingEpoch, this.nextDecisionRoot, {
+          state,
+          // have to use the "nextActiveIndices" that were saved in the last transition here to calculate
+          // the upcoming shuffling if it is not already built (similar condition to the below computation)
+          activeIndices: this.nextActiveIndices,
+        }) ??
+        // allow for this case during testing where the ShufflingCache is not present, may affect perf testing
+        // so should be taken into account when structuring tests.  Should not affect unit or other tests though
+        computeEpochShuffling(state, this.nextActiveIndices, upcomingEpoch);
+    }
+    const upcomingProposerSeed = getSeed(state, upcomingEpoch, DOMAIN_BEACON_PROPOSER);
+    // next epoch was moved to current epoch so use current here
+    this.proposers = computeProposers(upcomingProposerSeed, this.currentShuffling, this.effectiveBalanceIncrements);
+
+    // calculate next values
+    this.nextDecisionRoot = getShufflingDecisionBlock(state.config, state, epochAfterUpcoming);
+    this.nextActiveIndices = new Array<number>(nextActiveIndicesLength);
+
+    if (nextActiveIndicesLength > epochTransitionCache.nextEpochShufflingActiveValidatorIndices.length) {
+      throw new Error(
+        `Invalid activeValidatorCount: ${nextActiveIndicesLength} > ${epochTransitionCache.nextEpochShufflingActiveValidatorIndices.length}`
+      );
+    }
+    // only the first `activeValidatorCount` elements are copied to `activeIndices`
+    for (let i = 0; i < nextActiveIndicesLength; i++) {
+      this.nextActiveIndices[i] = epochTransitionCache.nextEpochShufflingActiveValidatorIndices[i];
+    }
+
+    if (this.shufflingCache) {
+      this.nextShuffling = null;
+      this.shufflingCache?.build(epochAfterUpcoming, this.nextDecisionRoot, state, this.nextActiveIndices);
+    } else {
+      this.nextShuffling = computeEpochShuffling(state, this.nextActiveIndices, epochAfterUpcoming);
```

Comment:
add a comment to say this is for spec test, should always have a shufflingCache for PROD beacon node

Reply:
Updated